add support for python 3.13

maurerle opened the pull request:
This PR adds support for Python 3.13 and updates required dependencies to work with latest versions.

supersedes #67 
closes #66 

Generally, I like the approach to be open to new versions much more, and only provide minimum versions better, but the current approach is of course fine as well, if it keeps being maintained.

t-gross committed:
Merge branch 'develop' into 'master'

Add JOSS article

See merge request iek-3/shared-code/fine!390

maurerle committed:
update dependencies and supported python versions

maurerle commented:
@t-gross is there an option to add this, as this closes an issue we have with gurobipy?

maurerle commented:
@JohannesBehrens could you please take a look? It has been 5 month now that FINE is not usable with python 3.13...

phil-fzj commented:
hi @maurerle. Sorry for the delay and thank you for your PR!
I just tested this and there seem to be no issues in our pipeline.

Could you also change the version constraints in the 
- requirements_dev.yml
- requirements.yml

Then we can merge this!

maurerle commented:
Finally :)
Thanks @phil-fzj  - i just updated this PR by rebasing and added your suggested change :)

julian-belina committed:
Setup Github pipeline

julian-belina committed:
rename requirements.yml

julian-belina committed:
Add pypi pytest

julian-belina committed:
Change pypi installation

julian-belina committed:
Check only fine directory

JohannesBehrens commented:
Thanks @maurerle for the code update.
We are currently working on migrating the full fine repository (including our test pipelines) to github. Once this is done I will merge your Request today.

julian-belina committed:
Change fine

julian-belina committed:
Change

julian-belina committed:
remove dot

julian-belina committed:
Exclude examples

julian-belina committed:
Add slash

julian-belina committed:
Change code formatting syntax

julian-belina committed:
Change formatting

julian-belina committed:
Add pypi dev install

julian-belina committed:
Merge remote-tracking branch 'origin/develop' into 70_add_ci

julian-belina committed:
Update test_pull_dev.yml

julian-belina committed:
Add test push to jugit, add separation between environment setup and tests, add more precise library printing for pypi

julian-belina committed:
Add untracked changes

julian-belina committed:
weird commit

julian-belina committed:
Fix git push syntax

julian-belina committed:
Change branch name on push to dev, add conda channels explicitly

julian-belina committed:
Change source repo

julian-belina committed:
Trigger develop push to jugit only on push to develop, add test for python versions

julian-belina committed:
Change name requirements_file in python test script, add fail fast

julian-belina committed:
Move channel definition to github action

julian-belina committed:
Adjust formatting in requirements.yml to match meta.yml in conda forge, remove nodefaults from channel list to supress warning

julian-belina committed:
test on self hosted

julian-belina committed:
Remove test environment

julian-belina committed:
track system ressources

julian-belina committed:
Use old miniforge version

julian-belina committed:
Try windows versions

julian-belina committed:
Remove fail fast

julian-belina committed:
Merge branch 'develop' into 70_add_ci

julian-belina commented:
@MaurerLe, thank you for your efforts to improve fine. Unfortunately, there are still some unresolved issues.

I have set up a pipeline to test Fine for versions 3.10, 3.11, 3.12 and 3.13, for installation via PyPi and Conda Forge. Therefore, I have forked your branch [here](https://github.com/julian-belina/FINE), @maurerle. So i can insepct Github actions more easily.

The results can be found here: https://github.com/FZJ-IEK3-VSA/FINE/actions/runs/17839572203/job/50796565110.


1. It seems that, when installing the dependencies from PyPI, our examples yield different results to the expected values for all Python versions. The .ipynb files appear to have some kind of built-in test. I have not yet figured out how these are implemented.
2. Fine with Python 3.13 is not installable from Conda Forge due to the pandas <=0.1.14 constraint.


The constraints cannot be officially relaxed until both issues have been resolved. @LarsHadidi  has already started working on this. I hope we can combine our efforts soon. Otherwise, @JohannesBehrens and @phil-fzj , we need to take a look at this.

phil-fzj commented:
To 1) 
The test works as such that the notebooks outputs are checked against the stored values of cells tagged with _nbval-check-output_.
The issue is that the stored version of the notebook was probably run with numpy<2 while the pipeline runs with numpy>2. This results in slightly different formatted outputs:
e.g. examples/01_1node_Energy_System_Workflow/01_1node_Example.ipynb 
output for `data["Wind (onshore), operationRateMax"].sum()`:
Original notebook cell content: 
`"text/plain": [ "2300.4069071646272" ]`
Cell output with numpy>2:
`"text/plain": ["np.float64(2300.4069071646272)"]`

This causes the tests to fail despite the values actually being the same.
We could think about casting all the outputs to float explicitly, e.g.:
`float(data["Wind (onshore), operationRateMax"].sum())`
then the output is:
`"text/plain": [ "2300.4069071646272" ]`
regardless of the numpy version.
 
This way, tests should pass for both numpy 1&2

maurerle commented:
@julian-belina
1. the issues you describe are unrelated to this PR.
2. Your pipeline now runs for different OS types but always with python3.10
3. I did not find a `pandas <=0.1.14` constraint - did you mean geopandas with `<=0.14.4` could be bumped to 1.x eventually?

maurerle committed:
fix output of ipynb notebooks

the change to np.float64 comes from upgrading to numpy 2.x
validating the ipynb output is probably not done in user tests, only in CI and should therefore be safe to always assume numpy>=2

julian-belina commented:
@maurerle Sorry, there were some errors in my last post.

The correct pipeline can be found at this link: https://github.com/julian-belina/FINE/actions/runs/17863033778. This workflow differs from the standard pipeline, which is intended for testing new features rather than updating dependencies.

It would be great if you could reopen the pull request for this branch: https://github.com/FZJ-IEK3-VSA/FINE/tree/70_add_ci. This would allow me to run a different pipeline.

Secondly, I meant geopandas <=0.14.4 and not pandas <=0.14.4. While installing from PyPI works fine, it is not possible to create the environment from Conda Forge when using requirements-dev.yml.
This causes this pipeline to fail. 

@phil-fzj  It seems that @maurerle  has already implemented this. Once the other pipeline has passed, we can merge it.

maurerle committed:
update additional packages for conda to solve versions

Bumping such a lot of versions is required in this case, as inidividual packages were not updated for a long time.

maurerle commented:
@julian-belina 
I added working yml versions to my branch (you can rebase, cherry-pick or merge this into your branch as needed).
Various dependencies were in need of a version bump (and some still are).

The output you saw in the pipeline is just a symptom. The root cause is that other packages - like psutil, xarray and numpy were in need of a relaxation of the version requirements.

In general, I would either have open dependency versions and regularly check that everything works in CI (as done in https://github.com/assume-framework/assume) or add a dependabot or whatever to get informed of new versions and incrementally test these.

julian-belina committed:
Merge branch '70_add_ci' into python_version

julian-belina committed:
Fix tests and examples

julian-belina committed:
Change tests

julian-belina committed:
Close issues

julian-belina reviewed (approved):
It passes all relevant tests and allows the installation of Python 3.10–3.13.

julian-belina commented:
@maurerle Thank you for your contribution.

Some adaptations to the tests were necessary to set the names of Pandas data frame indices. Interestingly, this only occurred when installing Pandas from Conda Forge. 

Unfortunately, due to a long backlog of maintenance tasks, we did not update the dependency requirements earlier. This will improve in future releases. However, we want to stick to constrained installation requirements so that fine can always be easily installed. Breaking changes in dependency releases occur quite often, and we cannot guarantee that we can immediately update to be compatible with them, but we try to keep up with the pace.

julian-belina commented on pyproject.toml:
```diff
@@ -34,22 +34,22 @@ dependencies = [
     "matplotlib<=3.9.2",
     "xlrd<=2.0.1",
-    "pyomo<=6.8.0",
-    "numpy<=1.26.4",
+    "pyomo<=6.9.4",
+    "numpy<=2.2.4",
     "pandas>=2,<=2.2.3",
-    "scipy<=1.14.1",
```
I didnt see that, but I will setup a test pipeline for older numpy versions and relax the constaints

phil-fzj replied:
my bad, i already removed the comment you are replying to as i noticed that I was wrong.
everything should be fine.
